### PR TITLE
Feature/pm 88 testing tutorials

### LINKED ISSUE #84: Create simple tutorial on how to run the tests locally

### PR DESCRIPTION
**Pull request by Olavhaasie**
Closes PM-88 and PM-83
Fixes #84 

## Description
Add a tutorial of running tests locally and how to write unit tests in python and c++ with a tutorial.

## Changes
* Added `ros_test_tutorial` subproject with some python and c++ library code
* Added `testing.rst` page

Please read the tutorials 😇

### CODE REVIEW

File: doc/development/testing.rst [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+.. code::
+
+  # colcon must be run from your workspace root
+  cd ~/march_ws
+  colcon build
+  colcon test
+
+The ``build`` step also builds the automated tests, so it is important to run
+``build`` when you modify tests. This will run all the tests and will fail when
+tests fail. Colcon also gives the possibility to review the test results and
```
Comment (Roelemans) [inline doc/development/testing.rst]: Here it seems like "build" runs all tests and will fail when tests fail.

File: doc/development/testing.rst [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+specified packages.
+
+.. code::
+
+  colcon test --packages-select march_some_package march_some_other_package
+
+Running roslint
+^^^^^^^^^^^^^^^
+The march packages that contain are also checked for code style with
+`roslint <https://wiki.ros.org/roslint>`_. The ``roslint`` target should also
```
Comment (RutgerVanBeek) [inline doc/development/testing.rst]: packages that contain what?

File: doc/development/testing.rst [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+The unit tests are different for c++ and python projects. C++ projects use the
+`gtest <https://github.com/google/googletest>`_ framework, whereas Python projects use
+`unittest <http://pythontesting.net/framework/unittest/unittest-introduction/>`_.
+For node level tests, `rostest <https://wiki.ros.org/rostest>`_ is used.
+Rostest uses launch files to launch the nodes under test and the actual tests.
+
+The next sections will describe the process of writing tests. For this
+tutorial we will use the :codedir:`ros_test_tutorial package <development/ros_test_tutorial>`.
+This package already includes some c++ and python code to write tests for.
+
```
Comment (RutgerVanBeek) [inline doc/development/testing.rst]: It is nice to have C++ with capital letter everywhere.

File: doc/development/testing.rst [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        target_link_libraries(add_test ros_test_tutorial ${catkin_LIBRARIES})
+    endif()
+
+Here we tell ``cmake`` to build the test and use our library. Now when
+``colcon test`` is run from the workspace root you should see tests passing.
+
+Writing c++ node tests
+^^^^^^^^^^^^^^^^^^^^^^
+
+
```
Comment (RutgerVanBeek) [inline doc/development/testing.rst]: Are you going to do this in the future and is there already an issue for this?

File: doc/development/testing.rst [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        catkin_add_nosetests(test/MultiplyTest.py)
+        ...
+    endif()
+
+The tests can also be run with ``colcon test``.
+Also see the `ros wiki on writing unit tests for python <https://wiki.ros.org/unittest#Code-level_Python_Unit_Tests>`_.
+
+Writing python node tests
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
Comment (RutgerVanBeek) [inline doc/development/testing.rst:212]: Are you going to do this in the future and is there already an issue for this?

**Review by RutgerVanBeek — CHANGES_REQUESTED** [review]

File: doc/development/testing.rst [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+^^^^^^^^^^^^^^^
+The march packages that contain are also checked for code style with
+`roslint <https://wiki.ros.org/roslint>`_. The ``roslint`` target should also
+be run from the workspace root like so.
+
+.. code::
+
+  colcon build --cmake-target-skip-unavailable --cmake-target roslint
+
+This command will run the ``roslint`` target on any packages that use
```
Comment (RutgerVanBeek) [inline doc/development/testing.rst]: Add the option --event-handlers console_direct+ to see lint warnings

File: doc/development/testing.rst [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+a ``package.xml`` or ``CMakeLists.txt`` file.
+
+General guidelines to writing tests
+-----------------------------------
+Here are some rules of thumb to follow when writing tests:
+
+1. Write short understandable tests.
+2. Seperate setup code that is necessary for every test.
+3. Do not depend on writing and reading from external files.
+4. Do not depend on timing in your tests.
```
Comment (Roelemans) [inline doc/development/testing.rst:79]: 'Vo

File: doc/development/testing.rst [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    TEST(AddTest, addNegative)
+    {
+        Add add;
+        ASSERT_EQ(add.add(0, -1), -1);
+    }
+
+In order to build and run the unit tests we must add ``rosunit`` as test
+dependency to our package. So normally we would add the following to the
+``package.xml``. However, this has already been done in highlevel
+``package.xml``.
```
Comment (Roelemans) [inline doc/development/testing.rst]:
```suggestion
``package.xml`` for this repository, so this is not necessary in this example.
```

File: doc/development/testing.rst [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+For node level tests, `rostest <https://wiki.ros.org/rostest>`_ is used.
+Rostest uses launch files to launch the nodes under test and the actual tests.
+
+The next sections will describe the process of writing tests. For this
+tutorial we will use the :codedir:`ros_test_tutorial package <development/ros_test_tutorial>`.
+This package already includes some c++ and python code to write tests for.
+
+Writing c++ unit tests
+^^^^^^^^^^^^^^^^^^^^^^
+Tests are always located in the ``test/`` directory of a package. To create a
```
Comment (Roelemans) [inline doc/development/testing.rst]:
```suggestion
As an example we will create tests for the function Add.cpp that is located in this repository. Tests are always located in the ``test/`` directory of a package.  To create a
```

File: doc/development/testing.rst [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+.. code::
+
+  <test_depend>rosunit</test_depend>
+
+Next we must tell cmake which tests it has to build. So add the following to the ``CMakeLists.txt``.
+
+.. code::
+
+    if(CATKIN_ENABLE_TESTING)
+        catkin_add_gtest(add_test test/AddTest.cpp)
```
Comment (Roelemans) [inline doc/development/testing.rst:152]: I do not understand what "add_test" means

File: doc/development/testing.rst [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+        target_link_libraries(add_test ros_test_tutorial ${catkin_LIBRARIES})
+    endif()
+
+Here we tell ``cmake`` to build the test and use our library. Now when
+``colcon test`` is run from the workspace root you should see tests passing.
+
+Writing c++ node tests
```
Comment (Roelemans) [inline doc/development/testing.rst]: Was this not also part of the issue?

File: doc/development/testing.rst [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+^^^^^^^^^^^^^^^^^^^^^^
+
+
+Writing python unit tests
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Tests for python are also always located in the ``test/`` directory. We will
```
Comment (Roelemans) [inline doc/development/testing.rst]:
```suggestion
As an example we will create tests for the function Multiply.py that is located in this repository. Tests are also always located in the ``test/`` directory. We will
```

File: doc/development/testing.rst [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+    from ros_test_tutorial import multiply
+
+    class MultiplyTest(unittest.TestCase):
+
+        def test_multiply_one_and_one():
+            self.assertEqual(multiply(1, 1), 1)
+
+    if __name__ == '__main__':
+        rosunit.unitrun(PKG, 'test_multiply', MultiplyTest)
```
Comment (Roelemans) [inline doc/development/testing.rst:188]: What does 'test_multiply' mean? Is that just a string that is displayed? Also, how does rosunit.unitrun now which functions in MultiplyTest are testfunctions it should run? (Does it run all functions, or all that start with "test" or something)

File: doc/development/testing.rst [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+The actual tests are written as functions inside the ``MultiplyTest`` class.
+See if you can add more tests (and make them fail). See the `python unittest
+documentation <https://docs.python.org/3/library/unittest.html>`_ for more
+information on writing tests.
+
+Next we must tell ``cmake`` to build and how to run the tests. We do this by
```
Comment (Roelemans) [inline doc/development/testing.rst]: What do you mean "how to run don't you just tell it to build them?

File: doc/development/testing.rst [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+adding ``catkin_add_nosetests()`` inside our ``CATKIN_ENABLE_TESTING``.
+
+.. code::
+
+    if(CATKIN_ENABLE_TESTING)
+        ...
+        catkin_add_nosetests(test/MultiplyTest.py)
```
Comment (Roelemans) [inline doc/development/testing.rst:203]: So this command is for python what 
catkin_add_gtest(add_test test/AddTest.cpp)
target_link_libraries(add_test ros_test_tutorial ${catkin_LIBRARIES})
is for c++?

File: doc/development/testing.rst [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+        ...
+    endif()
+
+The tests can also be run with ``colcon test``.
```
Comment (Roelemans) [inline doc/development/testing.rst]:
```suggestion
After building, the tests can now be run with ``colcon test``.
```

**Review by Roelemans — CHANGES_REQUESTED** [review]
Nice job. A few comments and some questions

File: doc/development/testing.rst (hunk 4 above)
Comment (Olavhaasie) [reply doc/development/testing.rst]: Yes: PM-109

File: doc/development/testing.rst (hunk 5 above)
Comment (Olavhaasie) [reply doc/development/testing.rst:212]: PM-109

File: doc/development/testing.rst (hunk 15 above)
Comment (Olavhaasie) [reply doc/development/testing.rst:203]: Exactly

**Review by RutgerVanBeek — DISMISSED** [review]

File: doc/development/testing.rst [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+root you should see tests passing.
+
+Writing C++ node tests
+^^^^^^^^^^^^^^^^^^^^^^
+
+
+Writing python unit tests
+^^^^^^^^^^^^^^^^^^^^^^^^^
+As an example we will create tests for the function ``Multiply.py`` that is
+located in this repository. Tests are also always located in the ``test/``
```
Comment (Roelemans) [inline doc/development/testing.rst]:
```suggestion
located in this repository. Tests are always located in the ``test/``
```

**Review by Roelemans — APPROVED** [review]
Excellent job. This is very useful :)

**Review by RutgerVanBeek — APPROVED** [review]

**Pull request merged**